lib/providers/password.js

```diff
@@ -14,33 +15,53 @@ var errors = require('../errors');
  *                                  have 2fa enabled.
  */
 function PasswordProvider (client, options) {
+    Provider.apply(this, arguments);
     // Jar for storing the session cookie
     this.jar = request.jar();
+    // storage of the CSRF Token as given in the response from the server
+    this.csrfToken = null;
     this.auth = options;
-    this.client = client;
 }
 
+PasswordProvider.CSRF_TOKEN_LOCATION = 'x-csrf-token';
+PasswordProvider.INVALID_CSRF_CODE = 461;
+
 /**
  * Attempts to authenticate with the given details. Resolves
- * with user information iff correct.
+ * with user information if correct.
  * @return {Promise}
  */
 PasswordProvider.prototype.attempt = function () {
     return this.client.request('post', '/users/login', { form: this.auth }).then(function (res) {
         if (res.statusCode !== 200) {
             throw new errors.AuthenticationFailedError(res);
-        } else {
-            return res;
         }
-    });
+        this.csrfToken = res.headers[PasswordProvider.CSRF_TOKEN_LOCATION];
+        return res;
+    }.bind(this));
 };
 
 /**
  * Returns info to add to the client's request.
  * @private
  */
 PasswordProvider.prototype.getRequest = function () {
-    return { jar: this.jar };
+    var req = {
+        jar: this.jar,
+        headers: {},
+    };
+    req.headers[PasswordProvider.CSRF_TOKEN_LOCATION] = this.csrfToken;
+    return req;
+};
+
+PasswordProvider.prototype.handleResponseError = function (err) {
```

Comment:
Perhaps, rather than returning a boolean from this (which isn't super obvious behaviour), it optionally returns a promise. Then we can have it just call `this.client.request` and return from there. A bit less single-purpose structure than what it is currently.